beacon-chain/cache/block.go

```diff
@@ -0,0 +1,104 @@
+package cache
+
+import (
+	"errors"
+	"strconv"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// ErrNotAncestorCacheObj will be returned when a cache object is not a pointer to
+	// block ancestor cache obj.
+	ErrNotAncestorCacheObj = errors.New("object is not an ancestor object for cache")
+	// Metrics
+	ancestorBlockCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "ancestor_block_cache_miss",
+		Help: "The number of ancestor block requests that aren't present in the cache.",
+	})
+	ancestorBlockCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "ancestor_block_cache_hit",
+		Help: "The number of ancestor block requests that are present in the cache.",
+	})
+	ancestorBlockCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "ancestor_block_cache_size",
+		Help: "The number of ancestor blocks in the ancestorBlock cache",
+	})
+)
+
+// AncestorInfo defines the cached ancestor block object for height.
+type AncestorInfo struct {
+	Hash   []byte
+	Height uint64
+	Block  *pb.BeaconBlock
+}
+
+// AncestorBlockCache structs with 1 queue for looking up block ancestor by height.
+type AncestorBlockCache struct {
+	ancestorBlockCache *cache.FIFO
+	lock               sync.RWMutex
+}
+
+// heightKeyFn takes the string representation of the block hash + height as the key
+// for the ancestor of a given block (AncestorInfo).
+func heightKeyFn(obj interface{}) (string, error) {
+	aInfo, ok := obj.(*AncestorInfo)
+	if !ok {
+		return "", ErrNotAncestorCacheObj
+	}
+
+	return string(aInfo.Hash) + strconv.Itoa(int(aInfo.Height)), nil
+}
+
+// NewBlockAncestorCache creates a new block ancestor cache for storing/accessing block ancestor
+// from memory.
+func NewBlockAncestorCache() *AncestorBlockCache {
+	return &AncestorBlockCache{
+		ancestorBlockCache: cache.NewFIFO(heightKeyFn),
+	}
+}
+
+// AncestorBySlot fetches block's ancestor by height. Returns true with a
+// reference to the ancestor block, if exists. Otherwise returns false, nil.
+func (a *AncestorBlockCache) AncestorBySlot(blockHash []byte, height uint64) (*AncestorInfo, error) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	obj, exists, err := a.ancestorBlockCache.GetByKey(string(blockHash) + strconv.Itoa(int(height)))
```

Comment:
why was this used ? why cant the key just be the blockhash

Reply:
good question, think of the following case:

```
a1--- b (v0-v4 voted for b)
   \-------c (v5-v6 voted for c)
    \----------d (v7 voted for d)
```

when we aggregate votes across the latest target blocks (b c and d), we need to distinguish keys b/c/d + whatever slot a1 is. Keep in mind the hash we are using here is not a block ancestor hash (is target block hash) because we dont know what the block ancestor hash is.

**Similar design:** 
https://github.com/ethereum/research/blob/master/ghost/ghost.py#L32
**Definition of a block ancestor:**
Get the ancestor of ``block`` with slot number ``slot``; return ``None`` if not found.